workflow/artifacts/gcs/gcs.go

```diff
@@ -0,0 +1,129 @@
+package gcs
+
+import (
+	"cloud.google.com/go/storage"
+	"context"
+	"errors"
+	argoErrors "github.com/argoproj/argo/errors"
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"os"
+)
+
+type GCSArtifactDriver struct {
+	Context context.Context
+}
+
+func (gcsDriver *GCSArtifactDriver) newGcsClient() (client *storage.Client, err error) {
+	gcsDriver.Context = context.Background()
+	client, err = storage.NewClient(gcsDriver.Context)
+	if err != nil {
+		return nil, argoErrors.InternalWrapError(err)
+	}
+	return
+
+}
+
+func (gcsDriver *GCSArtifactDriver) saveToFile(inputArtifact *wfv1.Artifact, filePath string) error {
+
+	log.Infof("Loading from GCS (gs://%s/%s) to %s",
+		inputArtifact.GCS.Bucket, inputArtifact.GCS.Key, filePath)
+
+	stat, err := os.Stat(filePath)
+	if err == nil {
+		if stat.IsDir() {
+			return errors.New("output artifact path is a directory")
+		}
+	}
+
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+	}
+
+	outputFile, err := os.OpenFile(filePath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
```

Comment:
blocker: Don't you want to use `O_TRUNC` and truncate anything in the file if it exists? I assume you want the GCS artifact to overwrite any existing contents. Can remove `O_APPEND` in that case.

Reply:
yeah definitely

Reply:
not sure what i was thinking there, good catch